product_code/user_service/impl/src/main/scala/de/upb/cs/uc4/user/impl/UserServiceImpl.scala

```diff
@@ -95,62 +95,73 @@ class UserServiceImpl(
 
       val validationErrorsFuture = postMessageUser match {
         case postMessageStudent: PostMessageStudent =>
-          var studentValidationErrors = postMessageStudent.validate
-          val student = postMessageUser.getUser.asInstanceOf[Student]
-          getAll(Role.Student).map(_.map(_.asInstanceOf[Student].matriculationId).contains(student.matriculationId)).map {
-            matDuplicate =>
-              if (matDuplicate) {
-                studentValidationErrors :+= SimpleError("student.matriculationId", "MatriculationID already in use.")
-              }
-              studentValidationErrors
+          //For students we may encounter duplicate matriculationIDs
+          postMessageStudent.validate.flatMap { studentValidationErrorsImmutable =>
+
+            var studentValidationErrors = studentValidationErrorsImmutable
+            val student = postMessageUser.getUser.asInstanceOf[Student]
+            getAll(Role.Student).map(_.map(_.asInstanceOf[Student].matriculationId).contains(student.matriculationId)).map {
+              matDuplicate =>
+                if (matDuplicate) {
+                  studentValidationErrors :+= SimpleError("student.matriculationId", "MatriculationID already in use.")
+                }
+                studentValidationErrors
+            }
           }
+
         case _ =>
-          Future.successful(postMessageUser.validate)
+          //For other users we cannot encounter duplicate matriculationIDs
+          postMessageUser.validate
       }
 
-      validationErrorsFuture.flatMap {
-        valErrors =>
-          var validationErrors = valErrors
+      var validationErrors = try {
+        Await.result(validationErrorsFuture, 5.seconds)
+      }
+      catch {
+        case _: TimeoutException => throw UC4Exception.ValidationTimeout
+        case e: Exception        => throw e
```

Comment:
This needs to be changed into a Internal Server Error
```suggestion
        case e: Exception        => throw UC4Exception.InternalServerError("Validation Error", e.getMessage)
```